[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/cindyweng18/csc448hw3/blob/main/Assignment3.ipynb)

In [1]:
# Imports section
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
url = "https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv"
df = pd.read_csv(url)
# Output the first 15 rows of the data
df.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [3]:
# Display a summary of the table information (number of datapoints, etc.)
df.describe()

Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

## Part 2. Splitting the dataset

In [4]:
# Take the pandas dataset and split it into our features (X) and label (y)
label = df['Size nm^3']
features = df[["Temperature °C", "Mols KCL"]]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.1)

## Part 3. Perform a Linear Regression

In [5]:
# Use sklearn to train a model on the training set
lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample = np.array([[534, 698]])
print(lin_reg.predict(sample))

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = lin_reg.score(X_train, y_train)
print("Score: ", score)

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coef = lin_reg.coef_
y_int = lin_reg.intercept_
print("Coefficients: ", coef)
print("Y-Intercept: ", y_int)

[773932.87314584]
Score:  0.8632400835615257
Coefficients:  [ 891.61683964 1037.23701147]
Y-Intercept:  -426181.953230476


$$
h(x) = -426181.9532 + (891.6168 * x_{1}) + (1037.237 * x_{2})
$$

The Linear Regression model has a score of 86%. This score means that the model can correctly predict a sample datapoint with 86% accuracy. The score is evaluated by dividing the sum of squared residuals with the total sum of squares and subtracting 1 to this value. 

## Part 4. Use Cross Validation

In [6]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(lin_reg, features, label)

# Report on their finding and their significance
scores

array([0.83918826, 0.87051239, 0.85871066, 0.87202623, 0.84364641])

The scores from the Cross Validation show a range between 84% to 87%, these scores are the scores at each fold of the cross validation method which by default is five (5) holds, that is why we see five scores. At each fold of the cross validation process, a set of data is separated to be trained and the rest to be tested with until we have done this for each set for all of the data provided. The scores reported earlier are consistent with the score in Part 3, both are in the same range.

## Part 5. Using Polynomial Regression

In [7]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(2)
new_X = poly.fit_transform(features)

X_train, X_test, y_train, y_test = train_test_split(new_X, label, test_size=0.1)

lin = LinearRegression()
lin.fit(X_train, y_train)

# Report on the metrics and output the resultant equation as you did in Part 3.
score = lin.score(X_train, y_train)
print("Score: ", score)

coef = lin.coef_
y_int = lin.intercept_
print("Coefficients: ", coef)
print("Y-Intercept: ", y_int)

Score:  1.0
Coefficients:  [ 0.00000000e+00  1.20000000e+01 -1.34516300e-07 -1.83657534e-11
  2.00000000e+00  2.85714287e-02]
Y-Intercept:  1.8080114386975765e-05


$$
h(x) = 1.8080 + (0 * x_{1}) + (12 * x_{2}) + (0.0000001345 * x_{3}) + (0.00000000001837 * x_{3}) + (2 * x_{4}) + (0.02857 * x_{5})
$$

The Linear Regression model after using PolynomialFeatures library reports a score of 100%. After expanding the features to a degree of 2, the model is able to correctly predict a sample datapoint with a 100% accuracy.